Source/Modules/python.cxx

```diff
@@ -626,6 +641,10 @@ class PYTHON:public Language {
       Printf(f_runtime, "#define SWIGPYTHON_FASTPROXY\n");
     }
 
+    if (stable_abi) {
+      Printf(f_runtime, "#define Py_LIMITED_API 0x03040000\n");
+    }
```

Comment:
It'd be helpful if the user could override this value (at least it will become helpful once SWIG can make use of `PyUnicode_AsUTF8AndSize()` being added to the stable ABI in 3.10 and the non-deprecated buffer API being added in 3.11).

Given `-stable-abi` doesn't actually affect SWIG's code-gen, maybe we should just generate a wrapper file that supports both (which is pretty much what we already seem to do) and leave defining `Py_LIMITED_API` or not and the choice of its value to the user - they could define it on the compiler command line, or in their interface file via something like:

```
%pythonbegin %{
#define Py_LIMITED_API 0x030A0000
%}
```
```

Reply:
What about generating

```cpp
#ifndef Py_LIMITED_API
#define Py_LIMITED_API 0x03040000
#endif
```

? This would still work by default out of the box but would allow overriding it with a later value by people knowing what they're doing.

Reply:
We could do this, the only downside is that we can't easily test for different versions of Py_LIMITED_API. Some users may even want to a workable way to build for different abi versions easily by specifying command line options.

So how about the following two ways of turning on the stable ABI as follows:

Either of:
1. Via the command line, for example to specify minimum of 3.4:
`-limitedapi=0x03040000`. The minimum version can be omitted as follows:
`-limitedapi`, whereupon the the minimum version of Python 3 that SWIG
 currently supports will be specified. Syntax follows that of the only other optional value that I can find (`-D`). Note that the command line name is changed to more accurately reflect what the value it is for. The difference between 'stable abi' and 'limited api' is somewhat blurred to me.
2. Via %begin, for example to specify minimum of 3.4:
```
%begin %{
#define Py_LIMITED_API 0x03040000
%}
```

How come we are currently specifying 0x03040000 as the minimum and not just 3 or 0x03020000 (the first version that the stable abi was introduced) or 0x03030000 (the minimum version that SWIG currently supports for non-stable abi)?

Reply:
> We could do this, the only downside is that we can't easily test for different versions of Py_LIMITED_API.

Why can't we? It's as simple as enabling stable ABI for different Python versions. Whether we should do this, and if it's worth it, is another question. IMO no because stable ABI is most (only?) useful with the lowest supported version anyhow.

> Some users may even want to a workable way to build for different abi versions easily by specifying command line options.

I don't see any need for this in practice as per above but why not.
 
> So how about the following two ways of turning on the stable ABI as follows:
> 
> Either of:
> 
>     1. Via the command line, for example to specify minimum of 3.4:
>        `-limitedapi=0x03040000`. The minimum version can be omitted as follows:
>        `-limitedapi`, whereupon the the minimum version of Python 3 that SWIG
>        currently supports will be specified.

The latter part _really_ doesn't make any sense to me, this is just not how/what stable ABI is for. Not specifying anything should default to the lowest supported version.

> Syntax follows that of the only other optional value that I can find (`-D`). Note that the command line name is changed to more accurately reflect what the value it is for. The difference between 'stable abi' and 'limited api' is somewhat blurred to me.

I prefer "stable ABI" because it's more clear, but the idea is that only limited API is included in stable ABI.

>     2. Via %begin, for example to specify minimum of 3.4:
> 
> ```
> %begin %{
> #define Py_LIMITED_API 0x03040000
> %}
> ```

Again, why not, but specifying `-DPy_LIMITED_API=0x03040000` in the compiler flags still looks much simpler for me and this is what I'd personally prefer to use. Why deal with the idiosyncratic SWIG flags when you can just use the well known (and stable since 50+ years) compiler `-D` switch?

> How come we are currently specifying 0x03040000 as the minimum and not just 3 or 0x03020000 (the first version that the stable abi was introduced) or 0x03030000 (the minimum version that SWIG currently supports for non-stable abi)?

I think 3.4 is the first version in which stable ABI is actually usable, it didn't include some absolutely required parts before then. I also think it would be perfectly fine to bump up the minimum SWIG supported version to 3.4.

Reply:
Adding a SWIG command line option to specify a `Py_LIMITED_API` value to hard-code into the generated file doesn't seem a helpful approach to me either.  SWIG generates a Python C/C++ wrapper with C/C++ preprocessor conditionals such that it can build with `Py_LIMITED_API` undefined or with it defined to a value of the builders choice.

Reply:
I worked out that 3.4 minimum is needed for the limited api as the wrappers require `PyType_GetSlot`.

Reply:
Okay, so the message I'm getting is that using `-D` when compiling the wrappers can define the version for `Py_LIMITED_API` and providing ways for SWIG to define this is not a good idea. Given the `-stable-abi` just defines `Py_LIMITED_API` and doesn't affect code generation in any other way, it seems that there is no point to the `-stable-abi` option at all and so I'll drop it altogether. The documented ways to turn on the stable abi will then be to define `Py_LIMITED_API` during C/C++ code compilation using either `%begin` or `-D` option passed to the C/C++ compiler. 

All round this is much simpler and less complicated for SWIG. I can also modify the test-suite to test for different versions of the abi by adding support for `EXTRA_CXXFLAGS` in Github CI. We need to test different abi version as there is a major change as @ojwb points out in our code base for version 3.10 (for `PyUnicode_AsUTF8AndSize()` for more efficient string handling) which needs testing.

Reply:
Thanks, I hadn't realized this but if SWIG can indeed generate the same code with and without `-stable-abi` and just compile it differently depending on whether `Py_LIMITED_API` is defined, it can indeed simplify things.

I think chances of someone defining ``Py_LIMITED_API` by accident are nil, practically speaking, so the only drawback I see in doing it like this is that it won't be very discoverable (e.g. you won't see it in `swig -help` output), but provided it's mentioned in the manual, it's probably not a great deal.

Reply:
Implemented in ef0bbcd6ed1a4f4356b885168082219b121adb11.